### ISSUE TITLE
Remove use of rfc822 module in twisted/mail and docs/core

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7868]
|[<img alt="graingert's avatar" src="https://avatars.githubusercontent.com/u/413772?s=50" width="50" height="50">](https://github.com/graingert)| @graingert reported|
|-|-|
|Trac ID|trac#7868|
|Type|enhancement|
|Created|2015-04-24 10:26:36Z|
|Branch|https://github.com/twisted/twisted/tree/7868-rodrigc-rfc822|

The rfc822 module was deprecated since Python 2.3: https://docs.python.org/2/library/rfc822.html
In Python 3, the rfc822 module was completely removed.
Any code which uses the rfc822 module will fail due to import errors.

People are supposed to use the [email](https://docs.python.org/2/library/email.html#module-email) package instead.

<details><summary>Searchable metadata</summary>

```
trac-id__7868 7868
type__enhancement enhancement
reporter__graingert graingert
priority__normal normal
milestone__Python_3_x Python-3.x
branch__7868_rodrigc_rfc822 7868-rodrigc-rfc822
branch_author__ 
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__None None
time__1429871196456694 1429871196456694
changetime__1474198049187875 1474198049187875
version__None None
owner__GitHub__noreply_____ GitHub <noreply@...>

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

What is "localmail"?

Please file a separate ticket for each of these modules.  Please do not attach a patch to this ticket which ports all of them at once.  Please see [Plan/Python3](https://twistedmatrix.com/trac/wiki/Plan/Python3).

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_5"></a>|@rodrigc commented|
|-|-|

I am repurposing this ticket for removal of rfc822 module from twisted/mail. :)

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_6"></a>|@rodrigc commented|
|-|-|

See:
https://github.com/twisted/twisted/pull/162

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban commented|
|-|-|

I have created the branch and sent it to buildbot

waiting for coverage report.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban set owner to @rodrigc|
|-|-|

Many thanks for the changes.

Changes look good.

Here are my comments:

1. For generateBounce in testBounceFormat I don't see that failedAddress is checked.

2. The changes from imap4.py and scripts/mailmail.py have no coverage https://codecov.io/gh/twisted/twisted/src/7868-rodrigc-rfc822/twisted/mail/imap4.py https://codecov.io/gh/twisted/twisted/src/7868-rodrigc-rfc822/twisted/mail/scripts/mailmail.py

Please improve the coverage and resubmit.

I think that we need to update the python-coverage tox runner to detect the PR env var and report it.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_10"></a>|@rodrigc set owner to @adiroiban|
|-|-|

At https://github.com/twisted/twisted/pull/162 ,
I added a test for looking for failedAddress inside the message transcript,
since generateBounce() takes failedAddress as a parameter, but doesn't return it directly.

For the second item, I don't understand the codecov.io web pages or how to proceed.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban removed owner|
|-|-|

The codecov.io pages are just simple HTML report for the coverage.

Lines highlighted in red mean that those lines are never executed as part of our test suite... so for those lines we can have gross errors (other than syntax errors) and we would not know.

So basically to make sure that your changes work we need minimal tests (unit tests) to exercise these changes.

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_12"></a>|@rodrigc set owner to @adiroiban|
|-|-|

I still don't understand the codecov web page.
Specifically, I don't understand what parts of my change does **not** have coverage.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_13"></a>|@adiroiban removed owner|
|-|-|

Ok. So first there is this diff

https://codecov.io/gh/twisted/twisted/compare/trunk...7868-rodrigc-rfc822

which is the `coverage` link from the ticket links.

In this diff we can see that for example line 154 from mailman.py has a red market at the start... this mean that this line is never called in our test suite.

You can replace that line with `o.sender = bad.module.parseaddr(hdrs[1])[1]` and the test suite should still pass.


https://codecov.io/gh/twisted/twisted/compare/trunk...7868-rodrigc-rfc822#747769737465642F6D61696C2F736372697074732F6D61696C6D61696C2E7079-154

--------

I don't know how else I can better describe the code coverage quality metric.

Leaving this for review so maybe another review can make it clear why and how we need those lines covered.

--------

Sorry for the trouble.

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_14"></a>|@rodrigc commented|
|-|-|

Moved to:
https://github.com/twisted/twisted/pull/242

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_15"></a>|@adiroiban commented|
|-|-|

Missing coverage:

* https://codecov.io/gh/twisted/twisted/compare/trunk...7868-rodrigc-rfc822#747769737465642F6D61696C2F736372697074732F6D61696C6D61696C2E7079-156

* https://codecov.io/gh/twisted/twisted/compare/trunk...7868-rodrigc-rfc822#747769737465642F6D61696C2F696D6170342E7079-1785


We are now enforcing 100% patch coverage so this branch will not land as it is.

The changes look very good but it needs a few more tests.

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_16"></a>|@rodrigc commented|
|-|-|

Patch updated, now achieving 100% patch coverage.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_17"></a>|@adiroiban set owner to @rodrigc|
|-|-|

Thanks for the changes.

Major comment:

1. test_searchOn talks about return True but all tests are with self.assertFalse ... if the test pass, it looks like there is a bug in the implementation

When writing a test please always see that it fails so that we don't write tests which are wrong.

---------

Minor comment.

1. In the test docstring, it is stated that they return True on some condition, but it should also explicitly state that False is return in condition is not met. In this way is not ambigour whetere None or an exception should be raised in these conditions.

Please check the tests and resubmit for review.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_18"></a>|@rodrigc commented|
|-|-|

I'm going to put this back in the review pile.
Due to the previous codecov requirements, I was pushed to write unit tests to cover a lot of mail functionality.

I do believe that there are bugs in the mail module, but I want to focus just on removing rfc822 module,
and not fixing all the other bugs in the mail module.  For test_searchOn() I could not generate a test case which
could trigger the true case.  I think there are some bugs in the mail module in this area.

Also, I am going to ask that adiroiban stop reviewing this ticket, so I can get a second opinion from another developer.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)|@rodrigc removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_20"></a>|GitHub set owner to GitHub<br>GitHub set status to `closed`|
|-|-|

In changeset 5b0977fda20855cb9070318081337c377abfb7fe
```
#!CommitTicketReference repository="" revision="5b0977fda20855cb9070318081337c377abfb7fe"
Merge pull request #242 from twisted/7868-rodrigc-rfc822

Author: rodrigc
Reviewer: markrwilliams
Fixes: #7868

Replace use of deprecated rfc822 module with email package
```

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_21"></a>|@hawkowl commented|
|-|-|

[mass edit] Removing review from closed tickets.
